sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ProcessMessageEventArgs.cs

```diff
@@ -27,13 +27,25 @@ public class ProcessMessageEventArgs : EventArgs
         /// </summary>
         public CancellationToken CancellationToken { get; }
 
+        /// <summary>
+        /// An event that is raised when the message lock is lost. This event is only raised for the scope of the Process Message handler.
+        /// Once the handler returns, the event will not be raised. There are two cases in which this event can be raised:
+        ///     1) When the message lock has expired based on the <see cref="ServiceBusReceivedMessage.LockedUntil"/> property.
+        ///     2) When a non-transient exception occurs while attempting to renew the message lock.
+        /// </summary>
+        public event Func<MessageLockLostEventArgs, Task> MessageLockLostAsync;
```

Comment:
This concerns me a bit.  There's a good chance that people are going to use a lambda here and may accidentally donk up lifetimes due to introducing a closure.   Before we release this, I think we need to do some weird lambda tests and see what it does to memory use over time.

Reply:
That's a great point - the whole idea behind putting this on the ProcessMessage args was to avoid requiring users to maintain a dictionary of message -> cancellationTokenSource, but there could be a perf implication due to the allocations.

Reply:
Allocations can be avoided by defining static closures - will add a sample for that. The other concern is that users may not remove their callbacks - will also include that in the sample.

Reply:
I'm biased I know 😉 But for me this discussion highlights that the new approach has now potentially more edge cases that need to be documented which complicates things for the end users

Reply:
To be fair though this approach shines in the case the event is raised multiple times when further messages are fetched over the receive actions. It seems bettee from an extensibility standpoint to support that scenario in the future without having to expose a way to get more tokens which I originally spiked